spring-kafka/src/test/java/org/springframework/kafka/retrytopic/AsyncMonoFutureRetryTopicClassLevelIntegrationTests.java

```diff
@@ -0,0 +1,971 @@
+import org.apache.kafka.common.serialization.StringDeserializer;
+import org.apache.kafka.common.serialization.StringSerializer;
+import org.assertj.core.api.InstanceOfAssertFactories;
+import org.junit.jupiter.api.Test;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.kafka.annotation.DltHandler;
+import org.springframework.kafka.annotation.EnableKafka;
+import org.springframework.kafka.annotation.KafkaHandler;
+import org.springframework.kafka.annotation.KafkaListener;
+import org.springframework.kafka.annotation.PartitionOffset;
+import org.springframework.kafka.annotation.RetryableTopic;
+import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory;
+import org.springframework.kafka.config.KafkaListenerEndpointRegistry;
+import org.springframework.kafka.config.TopicBuilder;
+import org.springframework.kafka.core.ConsumerFactory;
+import org.springframework.kafka.core.DefaultKafkaConsumerFactory;
+import org.springframework.kafka.core.DefaultKafkaProducerFactory;
+import org.springframework.kafka.core.KafkaAdmin;
+import org.springframework.kafka.core.KafkaAdmin.NewTopics;
+import org.springframework.kafka.core.KafkaTemplate;
+import org.springframework.kafka.core.ProducerFactory;
+import org.springframework.kafka.listener.ConcurrentMessageListenerContainer;
+import org.springframework.kafka.listener.ContainerProperties;
+import org.springframework.kafka.listener.ContainerProperties.AckMode;
+import org.springframework.kafka.listener.KafkaListenerErrorHandler;
+import org.springframework.kafka.support.Acknowledgment;
+import org.springframework.kafka.support.KafkaHeaders;
+import org.springframework.kafka.test.EmbeddedKafkaBroker;
+import org.springframework.kafka.test.context.EmbeddedKafka;
+import org.springframework.kafka.test.utils.KafkaTestUtils;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.converter.CompositeMessageConverter;
+import org.springframework.messaging.converter.GenericMessageConverter;
+import org.springframework.messaging.converter.SmartMessageConverter;
+import org.springframework.messaging.handler.annotation.Header;
+import org.springframework.retry.annotation.Backoff;
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
+import org.springframework.test.annotation.DirtiesContext;
+import org.springframework.test.context.TestPropertySource;
+import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;
+
+import reactor.core.publisher.Mono;
+
+/**
+ * @author Sanghyeok An
+ * @since 3.3.0
+ */
+
+@SpringJUnitConfig
+@DirtiesContext
+@EmbeddedKafka
+@TestPropertySource(properties = { "five.attempts=5", "kafka.template=customKafkaTemplate"})
+public class AsyncMonoFutureRetryTopicClassLevelIntegrationTests {
```

Comment:
What is the difference between this class and the next `AsyncMonoRetryTopicScenarioTests`?
Same question for all pairs of classes you have introduced here.
I even wonder if we don't have duplication in coveredge: I mean some tests might just do exactly the same what was done before.
Like this one is fully covered by the `CompletableFuture` one.
From the `MessagingMessageListenerAdapter` logic both `CompletableFuture` and `Mono` call the same `asyncFailure()`. So, your new code really does not introduce anything else `Mono`-specific.

My main concern that these new tests are too long to understand what is going on and they are possible too slow according to those 10 seconds sleep everywhere.

Reply:
As you mentioned, yes, its test cases is same with the one with `CompletableFuture`. 
When i started to handle this PR, i concerned that the difference of `Mono<?>` and `CompletableFuture<?>` cause an unexpected situation. 
As i concerned about it, it became true. 

<img width="800" alt="image" src="https://github.com/user-attachments/assets/bf280efd-85b7-4b90-91eb-82ac9772f99b">

The one `if ~ else` code which you pointed is the problem that i can find via both `CompletableFuture<?>` and `Mono<?>`. 
```java
handleException(...);
```

If we simplify that code as one line, only `CompletableFuture` test cases will fail and `Mono` test cases will success.